### ISSUE TITLE
By qwen2.5-3b, the GRPO performance is better than SDAR on webshop

### ISSUE THREAD

**Issue by zaiquanyang** [ZJU-REAL/SDAR#43]
hi authors,

I run the grpo trainer (https://github.com/ZJU-REAL/SDAR/blob/master/examples/grpo_trainer/run_webshop_8gpu.sh) and sdar trainer (https://github.com/ZJU-REAL/SDAR/blob/master/examples/sdar_trainer/run_webshop_3b.sh) on webshop, and obtain the results as below.

It seems that the **success rate** of **sdar** is even lower than **grpo**. Could the author please explain the problems that occurred during my reproduction? If any exist. 

In the final, I give the two scripts about grpo and sdar. No changes were made to the core parameters.

<img width="2885" height="556" alt="Image" src="https://github.com/user-attachments/assets/cd219b27-0d67-4be8-88dc-b1491bcd6038" />

**GRPO trainer**
```shell

ENGINE=${1:-vllm}
# export VLLM_ATTENTION_BACKEND=XFORMERS
export JAVA_HOME=$CONDA_PREFIX
export JVM_PATH=$JAVA_HOME/lib/server/libjvm.so
export PATH=$JAVA_HOME/bin:$PATH
export _JAVA_OPTIONS="-XX:+UseSerialGC -Xss512k -Xmx1g"
export OPENBLAS_NUM_THREADS=1
export MKL_NUM_THREADS=1
export OMP_NUM_THREADS=1
export DATASETS_MAX_WORKERS=1
export VLLM_USE_V1=0
ulimit -u 65536
# export RAY_object_store_memory=80000000000

num_cpus_per_env_worker=0.1 # The CPU resource allocated for each environment worker. If you want to use less CPU resources, you can decrease this value.

train_data_size=16
val_data_size=128
group_size=8
experiment_name="grpo_qwen2.5_3b"
project_name="sdar_agent_webshop"

# Model path.
MODELS_ROOT=${MODELS_ROOT:-/mnt/workspace/yyzq/data/dLLM/model_zoo}
MODEL_PATH=${MODEL_PATH:-$MODELS_ROOT/Qwen2.5-3B-Instruct}

# Swanlab configuration.
export SWANLAB_MODE="${SWANLAB_MODE:-cloud}"
export SWANLAB_API_KEY="${SWANLAB_API_KEY:-9r7HaIRTJNuhYahBZ2gVb}"

DEFAULT_LOCAL_DIR=${DEFAULT_LOCAL_DIR:-/mnt/workspace/yyzq/data/agent_rl/SDAR/ckpt/$experiment_name}
export SWANLAB_LOG_DIR="${SWANLAB_LOG_DIR:-${DEFAULT_LOCAL_DIR}/swanlab}"
mkdir -p "$SWANLAB_LOG_DIR"

# We only use data preparation to indicate the modality and the data size.
python3 -m examples.data_preprocess.prepare \
    --mode 'text' \
    --train_data_size $train_data_size \
    --val_data_size $val_data_size

python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files=$HOME/data/verl-agent/text/train.parquet \
    data.val_files=$HOME/data/verl-agent/text/test.parquet \
    data.train_batch_size=$train_data_size \
    data.val_batch_size=$val_data_size \
    data.max_prompt_length=4096 \
    data.max_response_length=512 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.return_raw_chat=True \
    actor_rollout_ref.model.path=$MODEL_PATH \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=64 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.01 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.name=$ENGINE \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.enable_chunked_prefill=False \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=False \
    actor_rollout_ref.rollout.val_kwargs.temperature=0.4 \
    actor_rollout_ref.rollout.val_kwargs.do_sample=True \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.use_invalid_action_penalty=True \
    actor_rollout_ref.actor.invalid_action_penalty_coef=0.1 \
    algorithm.use_kl_in_reward=False \
    env.env_name=Webshop \
    env.seed=0 \
    env.max_steps=15 \
    env.rollout.n=$group_size \
    env.resources_per_worker.num_cpus=$num_cpus_per_env_worker \
    trainer.critic_warmup=0 \
    trainer.logger=['console','swanlab'] \
    trainer.project_name=${project_name} \
    trainer.experiment_name=$experiment_name \
    trainer.default_local_dir=$DEFAULT_LOCAL_DIR \
    trainer.n_gpus_per_node=8 \
    trainer.ray_wait_register_center_timeout=600 \
    trainer.nnodes=1 \
    trainer.save_freq=-1 \
    trainer.test_freq=5 \
    trainer.total_epochs=320 \
    trainer.val_before_train=False $@
```

SDAR trainer
```shell

ENGINE=${1:-vllm}
# export VLLM_ATTENTION_BACKEND=XFORMERS
export JAVA_HOME=$CONDA_PREFIX
export JVM_PATH=$JAVA_HOME/lib/server/libjvm.so
export PATH=$JAVA_HOME/bin:$PATH
export _JAVA_OPTIONS="-XX:+UseSerialGC -Xss512k -Xmx1g"
export OPENBLAS_NUM_THREADS=1
export MKL_NUM_THREADS=1
export OMP_NUM_THREADS=1
export DATASETS_MAX_WORKERS=1
export VLLM_USE_V1=0
ulimit -u 65536
# export RAY_object_store_memory=80000000000

num_cpus_per_env_worker=0.1 # The CPU resource allocated for each environment worker. If you want to use less CPU resources, you can decrease this value.

# SDAR hyperparameters
sdar_coef=0.01
gate_beta=5.0
skill_all=false

train_data_size=16
val_data_size=128
group_size=8
experiment_name="sdar_qwen2.5_3b_coef${sdar_coef}_beta${gate_beta}_skillall${skill_all}"
project_name="sdar_agent_webshop"

# Model path.
MODELS_ROOT=${MODELS_ROOT:-/mnt/workspace/yyzq/data/dLLM/model_zoo}
MODEL_PATH=${MODEL_PATH:-$MODELS_ROOT/Qwen2.5-3B-Instruct}

# Swanlab configuration.
export SWANLAB_MODE="${SWANLAB_MODE:-cloud}"
export SWANLAB_API_KEY="${SWANLAB_API_KEY:-9r7HaIRTJNuhYahBZ2gVb}"

DEFAULT_LOCAL_DIR=${DEFAULT_LOCAL_DIR:-/mnt/workspace/yyzq/data/agent_rl/SDAR/ckpt/$experiment_name}
export SWANLAB_LOG_DIR="${SWANLAB_LOG_DIR:-${DEFAULT_LOCAL_DIR}/swanlab}"
mkdir -p "$SWANLAB_LOG_DIR"

# We only use data preparation to indicate the modality and the data size.
python3 -m examples.data_preprocess.prepare \
    --mode 'text' \
    --train_data_size $train_data_size \
    --val_data_size $val_data_size

python3 -m verl.trainer.main_sdar \
    algorithm.adv_estimator=grpo \
    data.train_files=$HOME/data/verl-agent/text/train.parquet \
    data.val_files=$HOME/data/verl-agent/text/test.parquet \
    data.train_batch_size=$train_data_size \
    data.val_batch_size=$val_data_size \
    data.max_prompt_length=4096 \
    data.max_response_length=512 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.return_raw_chat=True \
    actor_rollout_ref.model.path=$MODEL_PATH \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=64 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=8 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.01 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.name=$ENGINE \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.6 \
    actor_rollout_ref.rollout.enable_chunked_prefill=False \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=False \
    actor_rollout_ref.rollout.val_kwargs.temperature=0.4 \
    actor_rollout_ref.rollout.val_kwargs.do_sample=True \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=16 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    actor_rollout_ref.actor.use_invalid_action_penalty=True \
    actor_rollout_ref.actor.invalid_action_penalty_coef=0.1 \
    algorithm.use_kl_in_reward=False \
    +algorithm.sdar.sdar_coef=$sdar_coef \
    +algorithm.sdar.gate_beta=$gate_beta \
    +algorithm.sdar.skills_dir=skills/webshop \
    +algorithm.sdar.skill_all=$skill_all \
    env.env_name=Webshop \
    env.seed=0 \
    env.max_steps=15 \
    env.rollout.n=$group_size \
    env.resources_per_worker.num_cpus=$num_cpus_per_env_worker \
    trainer.critic_warmup=0 \
    trainer.logger=['console','swanlab'] \
    trainer.project_name=${project_name} \
    trainer.experiment_name=$experiment_name \
    trainer.default_local_dir=$DEFAULT_LOCAL_DIR \
    trainer.n_gpus_per_node=8 \
    trainer.ray_wait_register_center_timeout=600 \
    trainer.nnodes=1 \
    trainer.save_freq=20 \
    trainer.test_freq=5 \
    trainer.total_epochs=320 \
    trainer.val_before_train=False $@
```

**Comment by TianJ9** [comment]
almost the same. waiting too👀what's your performance on ALFWorld?

**Comment by zaiquanyang** [comment]
@TianJ9 
the below is the results by **qwen2.5-3b on alfworld**, although it was in the unseen validation set, it still reveal the problem that is t**he proposed method truly more effective than GRPO, or were there any issues with the evaluation**?

<img width="2904" height="1189" alt="Image" src="https://github.com/user-attachments/assets/024cb1bb-6992-4d48-9b43-e96cc59241e0" />

---

Besides, I also show the results  by **qwen2.5-7b on webshop**, the grpo performs better than the reported one. 
If I make selective choices of certain steps, at 195-step, it achieves 82.8. I guess the results of the author's paper were also obtained in this way though I think this kind of comparison is completely unreliable from the perspective of research work..

<img width="2925" height="1526" alt="Image" src="https://github.com/user-attachments/assets/25c0afc6-d260-4f6c-b1f5-f308252a97e6" />

@lll6gg, Could you please explain the problems that occurred during my reproduction?

**Comment by lll6gg** [comment]
Hi, could you please use our newest code to test ALFWorld on ID domains? 
We didn't have the finding that GRPO is better than SDAR.
<img width="2630" height="1266" alt="Image" src="https://github.com/user-attachments/assets/395a8177-36b0-4391-89a8-0a8dc96aa0f3" />

And from the results by qwen2.5-7b on webshop, the improvement speed is faster than GRPO in initial steps (0~100 steps), demonstrating our advantage against the baseline. The final best result may derive from randomness.

**Comment by lx-Meteors** [comment]
> 差不多一样。也在等呢👀你在ALFWorld上的表现如何？

我表现一般，79分，和论文不一样，我也提了issue，你呢

**Comment by TianJ9** [comment]
@lx-Meteors 3b在alfworld上复现结果78.91，和你差不多。

**Comment by zaiquanyang** [comment]
@TianJ9 @lx-Meteors 可以贴一下看看训练曲线日志么

**Comment by lx-Meteors** [comment]
> [@TianJ9](https://github.com/TianJ9) [@lx-Meteors](https://github.com/lx-Meteors) 可以贴一下看看训练曲线日志么

我单开了一个issue，里面贴了我的训练曲线～ 麻烦可以换到那边看一下嘛
https://github.com/ZJU-REAL/SDAR/issues/51

**Comment by lll6gg** [comment]
感谢大家的关注～可能有以下一些问题：
1. 是否用了最新的代码：[issue #35](https://github.com/ZJU-REAL/SDAR/issues/35) 解决了Skill Loading的一些问题并能达到更好的效果（82.8）。
2. 训练波动性：我们观察到确实波动较大，但多测两次发现3B均超过80（但并未超过多少），7B超过85，我们会进一步更新avg@3的数据。
3. 可以尝试一些我们的后续方法，比如[SEED](https://github.com/jinyangwu/SEED)，[OPID](https://github.com/jinyangwu/OPID)，[UCOB](https://github.com/TU2021/UCOB)，以减少skill质量带来的不稳定性

**Comment by pica15** [comment]
> [@TianJ9](https://github.com/TianJ9) the below is the results by **qwen2.5-3b on alfworld**, although it was in the unseen validation set, it still reveal the problem that is t**he proposed method truly more effective than GRPO, or were there any issues with the evaluation**?
> 
> <img alt="Image" width="2000" height="819" src="https://private-user-images.githubusercontent.com/48671947/617317956-024cb1bb-6992-4d48-9b43-e96cc59241e0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODUyNDA4OTMsIm5iZiI6MTc4NTI0MDU5MywicGF0aCI6Ii80ODY3MTk0Ny82MTczMTc5NTYtMDI0Y2IxYmItNjk5Mi00ZDQ4LTliNDMtZTk2Y2M1OTI0MWUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzI4VDEyMDk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3YmVkNmJmNDcyNDUzM2Q4ZTczYmU5ZjEwNDMyZjAyN2M3OTc4MGI2YzdmZWEwNjgxY2RkMmJkYTNkYTRlNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.e-OjcY-SnSr8UiyMY5_z45Pf1Do8_jXrBhB4gpnlPc8">
> Besides, I also show the results by **qwen2.5-7b on webshop**, the grpo performs better than the reported one. If I make selective choices of certain steps, at 195-step, it achieves 82.8. I guess the results of the author's paper were also obtained in this way though I think this kind of comparison is completely unreliable from the perspective of research work..
> 
> <img alt="Image" width="2000" height="1043" src="https://private-user-images.githubusercontent.com/48671947/617315047-25c0afc6-d260-4f6c-b1f5-f308252a97e6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODUyNDA4OTMsIm5iZiI6MTc4NTI0MDU5MywicGF0aCI6Ii80ODY3MTk0Ny82MTczMTUwNDctMjVjMGFmYzYtZDI2MC00ZjZjLWIxZjUtZjMwODI1MmE5N2U2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzI4VDEyMDk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1OTk3YWRjYTc4MjBhMTJiN2JmNGNlNDFiMTIzZGVmZTM3YzcwNDBmOThjZDFmMmFjZDMwYTkxM2NjYzYxZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.6TQbauzIkY9JqaX9yOS1AZqWvyGH1p3pO37IJdpaIYw">
> [@lll6gg](https://github.com/lll6gg), Could you please explain the problems that occurred during my reproduction?

@zaiquanyang Why are my reproduction results on WebShop significantly higher than the results reported in the paper across different models and algorithms? Could you share the specific evaluation configuration details?
